File: test/lyubashevsky.js
// LWE-based idscheme

var IDscheme = require('../idscheme.js');
//-----------------------------------Global variables-----------------------------------
var A_matrix;
var glob_y,
	glob_w;
var glob_tilde_w,
	glob_tilde_y;
var glob_m, 		
	glob_q,
	glob_n;
var c1,
	c2,
	c3;

//--------------------------------------------------------------------------------------

function keyGeneration(n, m, q) {
	var amatrix = IDscheme.initMatrixRandom(m, n, q); //pk
	glob_m = m;
	glob_n = n;
	glob_q = q;
	
	var tilde_w = [];

	for (var i = 0; i < glob_m; i++) { 
		tilde_w[i] = IDscheme.randInt(2);
	}
	
	// w = A.w_tilde mod q
	var aw = IDscheme.vectorMultiplyMatrix(tilde_w, amatrix);
	//	print(aw);
	var w = IDscheme.modVector(aw, glob_q);
	//	print("W:");
	//	print(w);
	
	A_matrix = amatrix;
	glob_w = w;
	glob_tilde_w = tilde_w;
	// print(tilde_w);
}

// Prover: p1()
function p1(){ 
	var amatrix = A_matrix;
	var tilde_y = [];
	
	for (var i = 0; i < glob_m; i++) { 
		tilde_y[i] = IDscheme.randInt(5*glob_m-1);
	}
	glob_tilde_y = tilde_y;
	//print(glob_tilde_y);

	var ay = IDscheme.vectorMultiplyMatrix(tilde_y, amatrix);
	//	print(ay);
	var y = IDscheme.modVector(ay, glob_q);

	glob_y = y;
	//print(y);
}

// Verifier: v1()
function v1(){
	var randomM = Math.floor(Math.random() * (+10 - +1)) + +1; 
	if(randomM>=5){
		var random = 1;
	}else{
		var random = 0;
	}
	return random;
}

// Prover: p2()
function p2(ch){
	var resp;
	var yw = IDscheme.addVectors(glob_tilde_y,glob_tilde_w);
	//print(yw);
	var refuse;
	for(var i=0;i<yw.length;i++){
		if (yw[i]<1 && yw[i]>(5*glob_m-1)){
			refuse = 1; // z perpendicular; 0 - refuse
		}else{
			refuse = 0;
		}
	}
	//print("Refuse: "+refuse);
	//print(ch);
	if(ch==1 && refuse==1){
		resp = 0;	
	}else{
		resp = [];
		resp[0] = glob_tilde_y;
		resp[1] = IDscheme.scalarMultiplyVector(ch, glob_w);
		// print("Resp:");
		// print(resp[1]);
	}
	
	return resp;
}

// Verifier: v2()
function v2(ch,params){
	var d;
	function norm(arr){ //compute norm
		var elements = arr.length;
		var total = 0;
		for(var i=0; i<elements; i++){
			  total += Math.pow(arr[i],2);
		}
		return Math.sqrt(total);
	}

	function compare(arr1,arr2){ //compare function
		for(var i=0; i<glob_m; i++){
			if(arr1[i]!=arr2[i])
				return false;
		}
		return true;
	}

	var condition = 5*Math.pow(glob_m,1.5);
	var norm_z = norm(params[0]);
	//print("Norm: " +norm_z);
	var z = params[0];
	amatrix = A_matrix;	
	//print(z.length);
	//print(amatrix.length);
	var a_transpose = IDscheme.transpose(amatrix);
	var a_transpose = IDscheme.transpose(amatrix);
	var comp1 = IDscheme.vectorMultiplyMatrix(z, amatrix);
	var az = IDscheme.modVector(comp1, glob_q);
	var comp2 = IDscheme.addVectors(params[1], glob_y);
	//print("Az mod q:");
	//print(az);
	//print("cw + y:");
	//print(comp2);

	if( norm_z<=condition && compare(az,comp2)){ 
			print("Success!");
			d=1;
	}else{
			print("Failed!");
			d=0;
	}



}

//===============================================================================
//------------------------------------------- start --------------------------------------------------
var n = 608, 
	q = 1024, 
	m = 960;


function testidscheme() {
	print("Lyubashevsky's ID Scheme:");
	print("Parameters Set:");
	print("n = " + n);
	print("m = " + m);
	print("q = " + q);

	var t0 = new Date().getTime();
	keyGeneration(n, m, q);
	var t1 = new Date().getTime();


	print("Prover: computes some parameters.");
	var t2 = new Date().getTime();
	p1();
	var t3 = new Date().getTime();

	var ch;
	ch = v1();

	print("Verifier: sends a challenge: " + ch);
	print("Prover: if challenge is 1, prover produces message of failure, otherwise, it reveals some parameters to the Verifier.");
	print("Verifier: checks correctness of computations from information revealed by prover and accepts in case of success. ");

	//var t4 = new Date().getTime();
	var response = p2(ch);
	//print(response);
	if(response==0){
		print("Failed!");
	}else{
		var t5 = new Date().getTime();
		v2(ch, response);
		var t6 = new Date().getTime();	
	}
	

	print("Time required by functions in ms:");
	print("Key Generation: " + (t1-t0) + " milliseconds");
	print("Prover (Computations of commitments): " + (t3-t2) + " milliseconds");
	//print("Prover (Revealing some parameters): " + (t5-t4) + " milliseconds");
	print("Verifier (Checking commitments' correctness): " + (t6-t5) + " milliseconds");
}
//***********************************************************
function print(message) {
	//WScript.Echo(message);
	console.log(message);
}
testidscheme();

File: id-schemes-java/IDschemeLibraryTest/src/Lyubashevsky.java
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Arrays;
import java.util.Random;
import java.math.BigInteger; 
import java.nio.charset.StandardCharsets; 
import java.security.MessageDigest; 
import java.security.NoSuchAlgorithmException; 


public class Lyubashevsky {
	// LWE-based IDscheme

	//-----------------------------------Global variables-----------------------------------

	static int[][] A_matrix;
	static int[] glob_w;
	static int[] glob_tilde_w;
	static int[] glob_y;
	static int[] glob_tilde_y;
	static int glob_m;
	static int glob_n;
	static int glob_q;
	

	//--------------------------------------------------------------------------------------

	public static int[] modVector(int[] v, int q) {
		int[] result = new int[v.length];
		if (q <= 0) {
			System.out.println("Modulus is not positive!");
			//break;
		}
		for (int i = 0; i < v.length; i++) {
			v[i] %= q;
			if (v[i] < 0) {
				v[i] += q;
			}
		}
		return v;
	}
	
	//Multiplies a vector by a scalar, c = s*v
	public static int[] scalarMultiplyVector(int s, int[] v) {
		int[] c = new int[v.length];
		for (int i = 0; i < v.length; i++) {
			c[i] = s * v[i];
		}
		return c;
	}
			
	public static void keyGeneration(int n, int m, int q) {
		int[][] amatrix = IDscheme.initMatrixRandom(m, n, q); // A, n*m
		int[] tilde_w = new int[m]; 
		glob_m = m;
		glob_n = n;
		glob_q = q;
		A_matrix = amatrix;

		for (int i = 0; i < glob_m; i++) { 
			tilde_w[i] = IDscheme.randInt(2);
		}
		
		// w = A.w_tilde mod q
		int[] aw = IDscheme.vectorMultiplyMatrix(tilde_w, amatrix);

		int[] w = modVector(aw, glob_q);
		
		A_matrix = amatrix;
		glob_w = w;
		glob_tilde_w = tilde_w;		
				
	}

	// Prover: p1()
	public static void p1(){ // 
		int[][] amatrix = A_matrix;
		int[] tilde_y = new int[glob_m]; 
		
		for (int i = 0; i < glob_m; i++) { 
			tilde_y[i] = IDscheme.randInt(5*glob_m-1);
		}
		glob_tilde_y = tilde_y;

		int[] ay = IDscheme.vectorMultiplyMatrix(tilde_y, amatrix);
		int[] y = modVector(ay, glob_q);

		glob_y = y;
		
	}
    
	// Verifier: v1()
	public static int v1(){
		int randnum = IDscheme.getRandomInteger(2, 0);
		return randnum;
	}
	
	// Prover: p2()
	public static Object[] p2(int c){
		int ch = c;
		Object[] resp = new Object[2];

		
		int[] yw = IDscheme.addVectors(glob_tilde_y,glob_tilde_w);
		int refuse = 0;
		for(int i=0;i<yw.length;i++){
			if (yw[i]<0 && yw[i]>(5*glob_m-1)){
				refuse = 1; // z perpendicular; 0 - refuse
			}else{
				refuse = 0;
			}
		}
		
		//refuse=1;
		//System.out.print(refuse);
		if(ch==1 && refuse==1){
			System.out.print("Failed!");
			System.exit(0);
		}
		
		int[] cw = scalarMultiplyVector(ch, glob_w);
		resp[0] = glob_tilde_y;
		resp[1] = cw;
		
		return resp;
	}
	
	public static double norm(int[] arr){ //compute norm
		int elements = arr.length;
		int total = 0;
			
		for(int i=0; i<elements; i++){
				  total += Math.pow(arr[i],2);
			}
			return Math.sqrt(total);
		}
	
	public static boolean compare(int[] arr1,int[] arr2){ //compare function
		for(int i=0; i<glob_m; i++){
			if(arr1[i]!=arr2[i]) {
				//System.out.println(arr1[i] + " ve " + arr2[i]);
				return false;
			}
		}
		return true;
	}
	
	// Verifier: v2()
	public static void v2(int ch, Object[] params){
		Object[] resp = params;
		/*System.out.println(Arrays.toString(resp));
		System.out.println(resp[0]);
		System.out.println(Arrays.toString((int[])resp[1]));
*/
	//	int p1 = (int)resp[0];
		
	//	System.out.print(String.valueOf(p1));
		for(int i=0;i<resp.length; i++) {
	//	System.out.print((int)resp[i]);
		}
	
		int d;
		double condition = 5*(Math.pow(glob_m,1.5));
		double norm_z = norm((int[])params[0]);
	
		int[] z = (int[]) params[0];
		int[][] amatrix = A_matrix;	
		
		int[] comp1 = IDscheme.vectorMultiplyMatrix(z, amatrix);
		int[] az = modVector(comp1, glob_q);
		int[] comp2 = IDscheme.addVectors((int[])params[1], glob_y);
	/*	System.out.println("Az length:" + az.length);
		System.out.println("cw+y length:" + comp2.length);
		
		for(int i =0;i<az.length;i++) {
			System.out.println("Az: " + az[i] + "cw + y =: " + comp2[i]);
		}
	*/
		//System.out.println(Arrays.toString(az).equals(Arrays.toString(comp2)));
		boolean condition2 = Arrays.toString(az).equals(Arrays.toString(comp2));
		if( norm_z<=condition && (condition2)){ 
			System.out.println("Success!");
			d=1;
		}else{
			System.out.println("Failed!");
			d=0;
	}
		
	}

	//===============================================================================
	//------------------------------------------- start --------------------------------------------------

	public static void main(String[] args) {
	
		int n = 608, 
			q = 1024, 
			m = 960;
	
		System.out.println("LWE-based ID Scheme:");
		System.out.println("Parameters Set:");
	/*	System.out.println("n = " + n);
		System.out.println("m = " + m);
		System.out.println("q = " + q);
	*/	long t0 = System.nanoTime();
		keyGeneration(n,m, q);
		long t1 = System.nanoTime();

	//	System.out.println("Prover: compute commitments c1, c2 and c3:");
		long t2 = System.nanoTime();
		p1();
		long t3 = System.nanoTime();

		int ch;
		ch = v1();
	//	System.out.println("Verifier: sends a challenge: " + ch);
	//	System.out.println("Prover: reveals some parameters depending on challenge.(Current challenge is: " + ch + ")");
		System.out.println("Verifier: checks commitment correctness from information revealed by prover and accept in case of success. ");
		long t4 = System.nanoTime();
		Object[] response = p2(ch);
		long t5 = System.nanoTime();
		v2(ch,response);
		long t6 = System.nanoTime();
		
		System.out.println("Keygen Execution time in milliseconds : " + 
				(t1-t0) / 1000000);
		System.out.println("Commitments comp Execution time in milliseconds : " + 
				(t3-t2) / 1000000);
		//System.out.println("Revealing some params Execution time in milliseconds : " + 
		//		(t5-t4) / 1000000);
		System.out.println("Verification Execution time in milliseconds : " + 
				(t6-t5) / 1000000);

	}

}
